File: src/Bindable/ObjectModel/CSharp/SodaFlow.Bindable.ObjectModel/BindableFactory.cs
using System;
using System.Collections.Generic;
using JetBrains.Annotations;
using SodaFlow.Functional;

namespace SodaFlow.Bindable.ObjectModel;

/// <summary>
///     The default <see cref="IBindableFactory" />. It holds one scheduler and gives that
///     scheduler to each object that it creates. Thus, a view model can take the factory through
///     its constructor, and a test can supply <see cref="BindingScheduler.Immediate" /> without a copy
///     of the true scheduler.
/// </summary>
[PublicAPI]
// ReSharper disable once InheritdocConsiderUsage
public class BindableFactory : IBindableFactory
{
    private readonly IBindingScheduler bindingScheduler;

    /// <summary>
    ///     Creates an instance that builds bindable objects with the given scheduler.
    /// </summary>
    /// <param name="bindingScheduler">
    ///     Moves notifications to the binding thread. An application captures it on the UI
    ///     thread at its start, with <see cref="SynchronizationContextBindingScheduler.Capture" />.
    ///     A test supplies <see cref="BindingScheduler.Immediate" />.
    /// </param>
    /// <exception cref="ArgumentNullException"><paramref name="bindingScheduler" /> is null.</exception>
    public BindableFactory(IBindingScheduler bindingScheduler) =>
        this.bindingScheduler = bindingScheduler ?? throw new ArgumentNullException(nameof(bindingScheduler));

    /// <inheritdoc />
    public IOneWayBindableValue<T> CreateOneWay<T>(Cell<T> cell, IEqualityComparer<T>? comparer = null) =>
        cell.ToOneWayImpl(scheduler: this.bindingScheduler, comparer: comparer);

    /// <inheritdoc />
    public ITwoWayBindableValue<T> CreateTwoWay<T>(
        Cell<T> cell,
        StreamSink<T> editsStreamSink,
        IEqualityComparer<T>? comparer = null) =>
        cell.ToTwoWayImpl(editsStreamSink: editsStreamSink, scheduler: this.bindingScheduler, comparer: comparer);

    /// <inheritdoc />
    public ITwoWayBindableValue<T> CreateTwoWay<T>(CellSink<T> sink, IEqualityComparer<T>? comparer = null) =>
        sink.ToTwoWayImpl(scheduler: this.bindingScheduler, comparer: comparer);

    /// <inheritdoc />
    public IOneWayToSourceBindableValue<T> CreateOneWayToSource<T>(
        StreamSink<T> editsStreamSink,
        T initialValue,
        IEqualityComparer<T>? comparer = null) =>
        editsStreamSink.ToOneWayToSourceImpl(
            initialValue: initialValue,
            scheduler: this.bindingScheduler,
            comparer: comparer);

    /// <inheritdoc />
    public IOneWayToSourceBindableValue<T> CreateOneWayToSource<T>(
        CellSink<T> sink,
        IEqualityComparer<T>? comparer = null) =>
        sink.ToOneWayToSourceImpl(scheduler: this.bindingScheduler, comparer: comparer);

    /// <inheritdoc />
    public IBindableAction<T> CreateBindableAction<T>(
        StreamSink<T> firingsStreamSink,
        Cell<bool>? isEnabledCell = null)
        where T : notnull =>
        firingsStreamSink.ToBindableActionImpl(scheduler: this.bindingScheduler, isEnabledCell: isEnabledCell);

    /// <inheritdoc />
    public IBindableAction CreateBindableAction(StreamSink<Unit> firingsStreamSink, Cell<bool>? isEnabledCell = null) =>
        new BindableActions.BindableAction(
            firingsStreamSink: firingsStreamSink,
            isEnabledCell: isEnabledCell,
            scheduler: this.bindingScheduler);

    /// <inheritdoc />
    public IBindableAction<Maybe<T>> CreateBindableAction<T>(
        StreamSink<Maybe<T>> firingsStreamSink,
        Cell<bool>? isEnabledCell = null)
        where T : notnull =>
        new BindableActions.BindableMaybeAction<T>(
            firingsStreamSink: firingsStreamSink,
            isEnabledCell: isEnabledCell,
            scheduler: this.bindingScheduler);
}

File: src/Bindable/ObjectModel/FSharp/SodaFlow.FSharp.Bindable.ObjectModel/BindableFactory.fs
namespace SodaFlow.Bindable.ObjectModel

open System
open System.Collections.Generic
open SodaFlow

type IBindableFactory =
    abstract ToOneWay<'T> : cell: Cell<'T> * ?comparer: IEqualityComparer<'T> -> IOneWayBindableValue<'T>

    abstract ToTwoWay<'T> :
        cell: Cell<'T> * editsStreamSink: StreamSink<'T> * ?comparer: IEqualityComparer<'T> -> ITwoWayBindableValue<'T>

    abstract ToTwoWay<'T> : sink: CellSink<'T> * ?comparer: IEqualityComparer<'T> -> ITwoWayBindableValue<'T>

    abstract ToOneWayToSource<'T> :
        editsStreamSink: StreamSink<'T> * initialValue: 'T * ?comparer: IEqualityComparer<'T> ->
            IOneWayToSourceBindableValue<'T>

    abstract ToOneWayToSource<'T> :
        sink: CellSink<'T> * ?comparer: IEqualityComparer<'T> -> IOneWayToSourceBindableValue<'T>

    abstract ToBindableAction<'T> :
        firingsStreamSink: StreamSink<'T> * ?isEnabledCell: Cell<bool> -> IBindableAction<'T>

    abstract ToBindableAction: firingsStreamSink: StreamSink<unit> * ?isEnabledCell: Cell<bool> -> IBindableAction

    /// A command whose CommandParameter can be null, a 'T, or a 'T option.
    abstract ToBindableOptionAction<'T> :
        firingsStreamSink: StreamSink<'T option> * ?isEnabledCell: Cell<bool> -> IBindableAction<'T option>

/// The one way to build a bindable. It holds one scheduler and gives that scheduler to each
/// object that it creates. An application captures the scheduler on the UI thread at its start,
/// with SynchronizationContextBindingScheduler.Capture, and gives the factory to each view model.
/// A test supplies BindingScheduler.Immediate.
type BindableFactory(scheduler: IBindingScheduler) =
    do
        if isNull (box scheduler) then
            raise (ArgumentNullException(nameof scheduler))

    interface IBindableFactory with
        member this.ToOneWay(cell, comparer) =
            BindableCoreExtensionMethods.ToOneWayImpl(cell, scheduler, comparer |> Option.defaultValue null)

        member this.ToTwoWay(cell, editsStreamSink, comparer) =
            BindableCoreExtensionMethods.ToTwoWayImpl(
                cell,
                editsStreamSink,
                scheduler,
                comparer |> Option.defaultValue null
            )

        member this.ToTwoWay(sink, comparer) =
            BindableCoreExtensionMethods.ToTwoWayImpl(sink, scheduler, comparer |> Option.defaultValue null)

        member this.ToOneWayToSource(editsStreamSink, initialValue, comparer) =
            BindableCoreExtensionMethods.ToOneWayToSourceImpl(
                editsStreamSink,
                initialValue,
                scheduler,
                comparer |> Option.defaultValue null
            )

        member this.ToOneWayToSource(sink, comparer) =
            BindableCoreExtensionMethods.ToOneWayToSourceImpl(sink, scheduler, comparer |> Option.defaultValue null)

        member this.ToBindableAction(firingsStreamSink, isEnabledCell: Cell<bool> option) =
            BindableCoreExtensionMethods.ToBindableActionImpl(
                firingsStreamSink,
                scheduler,
                isEnabledCell |> Option.defaultValue null
            )

        member this.ToBindableAction(firingsStreamSink, isEnabledCell: Cell<bool> option) : IBindableAction =
            new Bindable.BindableAction(firingsStreamSink, isEnabledCell, scheduler)

        member this.ToBindableOptionAction(firingsStreamSink, isEnabledCell: Cell<bool> option) =
            new Bindable.BindableOptionAction<_>(firingsStreamSink, isEnabledCell, scheduler)
